src/corehost/cli/fxr/fx_muxer.cpp

```diff
@@ -26,20 +29,52 @@
 #include "roll_fwd_on_no_candidate_fx_option.h"
 
 using corehost_main_fn = int(*) (const int argc, const pal::char_t* argv[]);
-using corehost_get_delegate_fn = int(*)(coreclr_delegate_type type, void** delegate);
 using corehost_main_with_output_buffer_fn = int(*) (const int argc, const pal::char_t* argv[], pal::char_t buffer[], int32_t buffer_size, int32_t* required_buffer_size);
 
+namespace
+{
+    // hostfxr tracks the context used to load hostpolicy and coreclr as the active host context. This is the first
+    // context that is successfully created and used to load the runtime. There can only be one active context.
+    // Secondary contexts can only be created once the first context has fully initialized and been used to load the
+    // runtime. Any calls to initialize another context while the runtime has not yet been loaded by the first context
+    // will block until the first context loads the runtime (or fails).
+    std::mutex g_context_lock;
+
+    // Tracks the active host context. This is the context that was used to load and initialize hostpolicy and coreclr.
+    // It will only be set once both hostpolicy and coreclr are loaded and initialized. Once set, it should not be changed.
+    std::unique_ptr<const host_context_t> g_active_host_context;
+
+    // Tracks whether the host context is initializing (from creation of the first context to loading the runtime).
+    // Initialization of other contexts should block if the first context is initializing (i.e. this is true).
+    // The condition variable is used to block on and signal changes to this state.
+    std::atomic<bool> g_context_initializing(false);
+    std::condition_variable g_context_cv;
+
+    void handle_initialize_failure_or_abort(const hostpolicy_contract_t *hostpolicy_contract = nullptr)
+    {
+        {
+            std::lock_guard<std::mutex> lock{ g_context_lock };
+            g_context_initializing.store(false);
```

Comment:
Please assert that `g_context_initializing` is true if we get here... 
And maybe also assert that there is no `g_active_host_context`.